RdcMan save code, rebuild with errors `error CS0205: Cannot call an abstract base member`

RDCMan Download:
https://www.microsoft.com/en-us/download/details.aspx?id=44989

US VS2015(C# 6) or 2019 (C# 8)  Rebuild with errors: 
```
1>RDCMan\Properties\AssemblyInfo.cs(23,31,23,60): warning CS0618: 'SecurityAction.RequestMinimum' is obsolete: 'Assembly level declarative security is obsolete and is no longer enforced by the CLR by default. See http://go.microsoft.com/fwlink/?LinkID=155570 for more information.'
1>RdcMan\Server.cs(91,48,91,63): error CS0205: Cannot call an abstract base member: 'ServerBase.Properties'
1>RdcMan\Server.cs(93,60,93,80): error CS0205: Cannot call an abstract base member: 'ServerBase.DisplaySettings'
1>RdcMan\Program.cs(231,35,231,51): error CS7069: Reference to type 'ReflectionContext' claims it is defined in 'mscorlib', but it could not be found
1>RdcMan\Server.cs(474,4,474,19): error CS0200: Property or indexer 'ServerBase.Properties' cannot be assigned to -- it is read only
1>RdcMan\Server.cs(475,4,475,24): error CS0200: Property or indexer 'ServerBase.DisplaySettings' cannot be assigned to -- it is read only
1>RdcMan\ListSessionsForm.cs(14,22,14,32): warning CS0649: Field 'ListSessionsForm.components' is never assigned to, and will always have its default value null
```

--- a/ICSharpCode.Decompiler.Tests/TestCases/Pretty/QualifierTests.cs
+++ b/ICSharpCode.Decompiler.Tests/TestCases/Pretty/QualifierTests.cs
@@ -155,6 +155,49 @@ namespace ICSharpCode.Decompiler.Tests.Pretty
 			}
 		}
 
+		public class Root
+		{
+			private int prop;
+
+#if LEGACY_CSC
+			public int Prop {
+				get {
+					return prop;
+				}
+			}
+#else
+			public int Prop => prop;
+#endif
+			public void M<T>(T a)
+			{
+
+			}
+		}
+
+		public abstract class Base : Root
+		{
+			public new abstract int Prop { get; }
+
+			public new abstract void M<T>(T a);
+		}
+
+		public class Derived : Base
+		{
+#if LEGACY_CSC
+			public override int Prop {
+				get {
+					return ((Root)this).Prop;
+				}
+			}
+#else
+			public override int Prop => ((Root)this).Prop;
+#endif
+			public override void M<T>(T a)
+			{
+				((Root)this).M(a);
+			}
+		}
+
 		private int fieldConflict;
 		private int innerConflict;
 

--- a/ICSharpCode.Decompiler/CSharp/ExpressionBuilder.cs
+++ b/ICSharpCode.Decompiler/CSharp/ExpressionBuilder.cs
@@ -2487,9 +2487,11 @@ namespace ICSharpCode.Decompiler.CSharp
 			{
 				if (ShouldUseBaseReference())
 				{
+					var baseReferenceType = resolver.CurrentTypeDefinition.DirectBaseTypes
+						.FirstOrDefault(t => t.Kind != TypeKind.Interface);
 					return new BaseReferenceExpression()
 						.WithILInstruction(target)
-						.WithRR(new ThisResolveResult(memberDeclaringType, nonVirtualInvocation));
+						.WithRR(new ThisResolveResult(baseReferenceType ?? memberDeclaringType, nonVirtualInvocation));
 				}
 				else
 				{

--- a/ICSharpCode.Decompiler/CSharp/TranslatedExpression.cs
+++ b/ICSharpCode.Decompiler/CSharp/TranslatedExpression.cs
@@ -452,6 +452,7 @@ namespace ICSharpCode.Decompiler.CSharp
 				// perform remaining pointer cast, if necessary
 				return pointerExpr.ConvertTo(targetType, expressionBuilder);
 			}
+			Expression expr;
 			if (targetType.Kind == TypeKind.ByReference)
 			{
 				if (NormalizeTypeVisitor.TypeErasure.EquivalentTypes(targetType, this.Type))
@@ -481,7 +482,6 @@ namespace ICSharpCode.Decompiler.CSharp
 				// Convert from integer/pointer to reference.
 				// First, convert to the corresponding pointer type:
 				var arg = this.ConvertTo(new PointerType(elementType), expressionBuilder, checkForOverflow);
-				Expression expr;
 				ResolveResult elementRR;
 				if (arg.Expression is UnaryOperatorExpression unary && unary.Operator == UnaryOperatorType.AddressOf)
 				{
@@ -561,7 +561,11 @@ namespace ICSharpCode.Decompiler.CSharp
 					return this;
 				}
 			}
-			var castExpr = new CastExpression(expressionBuilder.ConvertType(targetType), Expression);
+			// BaseReferenceExpression must not be used with CastExpressions
+			expr = Expression is BaseReferenceExpression
+				? new ThisReferenceExpression().WithILInstruction(this.ILInstructions)
+				: Expression;
+			var castExpr = new CastExpression(expressionBuilder.ConvertType(targetType), expr);
 			bool needsCheckAnnotation = targetUType.GetStackType().IsIntegerType();
 			if (needsCheckAnnotation)
 			{

